pkg/endpoint/policy_test.go

```diff
@@ -38,3 +53,163 @@ func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
 	})
 	c.Assert(ep.visibilityPolicy, check.IsNil)
 }
+
+// This test fuzzes the incremental update engine from an end-to-end perspective
+// to ensure we don't ever miss an incremental update.
+//
+// It works by simulating a "churning" IPcache that is constantly allocating new identities.
+// There is a single policy that -- funnily enough -- selects all of the new identities.
+// We then continuously simulate endpoint regeneration and ensure the computed policy contains
+// all the generated identities.
+//
+// By default, we test 1000 identities, which should take less than 10 seconds. If this test fails,
+// please bump the factor to something massive and start debugging :-).
+func TestIncrementalUpdatesDuringPolicyGeneration(t *testing.T) {
+	const testfactor = 1000 // bump this to stress-test
+
+	policy.SetPolicyEnabled("always")
+
+	idcache := make(cache.IdentityCache, testfactor)
+	fakeAllocator := testidentity.NewMockIdentityAllocator(idcache)
+	repo := policy.NewPolicyRepository(fakeAllocator, fakeAllocator.GetIdentityCache(), nil, nil)
+
+	defer func() {
+		repo.RepositoryChangeQueue.Stop()
+		repo.RuleReactionQueue.Stop()
+		repo.RepositoryChangeQueue.WaitToBeDrained()
+		repo.RuleReactionQueue.WaitToBeDrained()
+	}()
+
+	addIdentity := func(labelKeys ...string) *identity.Identity {
+		t.Helper()
+		lbls := labels.Labels{}
+		for _, labelKey := range labelKeys {
+			lbls[labelKey] = labels.NewLabel("k8s:"+labelKey, "", "")
+		}
+		id, _, err := fakeAllocator.AllocateIdentity(context.Background(), lbls, false, 0)
+		if err != nil {
+			t.Fatal(err)
+		}
+		//t.Logf("allocated label %s id %d", labelKeys, id.ID) // commented out for speed
+
+		wg := &sync.WaitGroup{}
+		repo.GetSelectorCache().UpdateIdentities(cache.IdentityCache{
+			id.ID: id.LabelArray,
+		}, nil, wg)
+		wg.Wait()
+		return id
+	}
+
+	podID := addIdentity("pod")
+	repo.GetPolicyCache().LocalEndpointIdentityAdded(podID)
+
+	ep := Endpoint{
+		SecurityIdentity: podID,
+		policyGetter:     &mockPolicyGetter{repo},
+	}
+	ep.UpdateLogger(nil)
+
+	podSelectLabel := labels.ParseSelectLabel("pod")
+	egressSelectLabel := labels.ParseSelectLabel("peer")
+
+	// Create a rule that selects our pod, and allows all peer identities
+	egressDenyRule := &api.Rule{
+		EndpointSelector: api.NewESFromLabels(podSelectLabel),
+		EgressDeny: []api.EgressDenyRule{
```

Comment:
How does this allow all peer identities if it's an egress deny rule...?

Reply:
it's a typo - the word should be "selects", not allows. I had used allow rules but it got coalesced in to a wildcard, neat.